src/devices/src/virtio/queue.rs

```diff
@@ -296,6 +303,32 @@ impl Queue {
             return None;
         }
 
+        self.do_pop(mem)
+    }
+
+    /// Try to pop the first available descriptor chain from the avail ring.
+    /// If no descriptor is available, enable notifications.
+    pub fn pop_or_enable_notification<'a, 'b>(
```

Comment:
Do we want to have a future proof approach here so that we end up in an ideal state of being able to quickly switch to rust-vmm implementation ? https://github.com/rust-vmm/vm-virtio/blob/master/crates/virtio-queue/src/lib.rs

Reply:
Yes, of course. Actually this implementation is my interpretation of an idea suggested by @alexandruag who works closely with the rust-vmm community. I'm not sure how easy it would be to implement this in an iterator, but this way of enabling notifications simplifies the devices implementation by a lot. Otherwise in the devices we would have something like:

```
outter': loop {
    while let Some(chain) = queue.pop() {
        if rate_limit(chain) {
            break outter;
        }

        process(chain);
    }

    if !queue.enable_notification() {
        break;
    }
}
``` 

Especially in the net device this kind of implementation would be really hard to follow and I think it would lead to complications.

Reply:
As we discussed offline, this shouldn't break compatibility with rust-vmm.